pandas/tests/groupby/test_apply.py

```diff
@@ -1331,3 +1331,43 @@ def test_result_name_when_one_group(name):
     expected = Series([1, 2], name=name)
 
     tm.assert_series_equal(result, expected)
+
+
+def test_empty_df():
+    # GH 47985
+    empty_df = DataFrame({"a": [], "b": []})
+
+    # Both operations should return an empty series instead of IndexError for apply UDF
+    result1 = empty_df.groupby("a", group_keys=True).b.apply(lambda x: x.values[-1])
+    result2 = empty_df.groupby("a", group_keys=True).b.agg("sum")
+
+    expected = Series(
+        [], name="b", dtype="float64", index=Index([], dtype="float64", name="a")
+    )
+
+    tm.assert_series_equal(result1, expected)
+    tm.assert_series_equal(result2, expected)
+
+
+@pytest.mark.parametrize(
+    "error_type",
+    [
+        TypeError,
+        ValueError,
+        IndexError,
+    ],
+)
+def test_udf_raise_error_on_empty_df(error_type):
+    # GH 47985
+    empty_df = DataFrame({"a": [], "b": []})
+
+    def f(group):
+        raise error_type
+
+    # Exception should not be raised.
```

Comment:
I would expect the exception in `f` to actually raise in this case.

Since this isn't directly related the original issue, maybe this can be addressed later.